events/event_queue.go

```diff
@@ -0,0 +1,174 @@
+	EventCheckFailed = "check.failed"
+
+	EventIncidentCreated             = "incident.created"
+	EventIncidentResponderAdded      = "incident.responder.added"
+	EventIncidentResponderRemoved    = "incident.responder.removed"
+	EventIncidentCommentAdded        = "incident.comment.added"
+	EventIncidentDODAdded            = "incident.dod.added"
+	EventIncidentDODPassed           = "incident.dod.passed"
+	EventIncidentDODRegressed        = "incident.dod.regressed"
+	EventIncidentStatusOpen          = "incident.status.open"
+	EventIncidentStatusClosed        = "incident.status.closed"
+	EventIncidentStatusMitigated     = "incident.status.mitigated"
+	EventIncidentStatusResolved      = "incident.status.resolved"
+	EventIncidentStatusInvestigating = "incident.status.investigating"
+	EventIncidentStatusCancelled     = "incident.status.cancelled"
+
+	EventPushQueueCreate = "push_queue.create"
+)
+
+type Config struct {
+	UpstreamPush upstream.UpstreamConfig
+}
+
+// consumerWatchEvents keeps a registry of all the event_queue consumer and the events they watch.
+// This helps in ensuring that a single event is not being consumed by multiple consumers.
+var consumerWatchEvents = map[string][]string{
+	"team": {
+		EventTeamUpdate,
+		EventTeamDelete,
+	},
+	"notification": {
+		EventNotificationUpdate, EventNotificationDelete,
+		EventIncidentCreated,
+		EventIncidentResponderRemoved,
+		EventIncidentDODAdded, EventIncidentDODPassed, EventIncidentDODRegressed,
+		EventIncidentStatusOpen, EventIncidentStatusClosed, EventIncidentStatusMitigated,
+		EventIncidentStatusResolved, EventIncidentStatusInvestigating, EventIncidentStatusCancelled,
+		EventCheckPassed, EventCheckFailed,
+	},
+	"notification_send": {
+		EventNotificationSend,
+	},
+	"responder": {
+		EventIncidentResponderAdded,
+		EventIncidentCommentAdded,
+	},
+	"push_queue": {
+		EventPushQueueCreate,
+	},
+}
+
+func StartConsumers(gormDB *gorm.DB, pgpool *pgxpool.Pool, config Config) {
+	uniqWatchEvents := set.New[string]()
+	for _, v := range consumerWatchEvents {
+		for _, e := range v {
+			if uniqWatchEvents.Contains(e) {
+				logger.Fatalf("Error starting consumers: event[%s] has multiple consumers", e)
+			}
+
+			uniqWatchEvents.Add(e)
+		}
+	}
+
+	allConsumers := []*EventConsumer{
+		NewTeamConsumer(gormDB, pgpool),
+		NewNotificationConsumer(gormDB, pgpool),
+		NewNotificationSendConsumer(gormDB, pgpool),
+		NewResponderConsumer(gormDB, pgpool),
+	}
+	if config.UpstreamPush.Valid() {
+		allConsumers = append(allConsumers, NewUpstreamPushConsumer(gormDB, pgpool, config))
+	}
+
+	for i := range allConsumers {
+		go allConsumers[i].Listen()
+	}
+}
+
+// newEventQueueConsumerFunc returns a new event consumer for the `event_queue` table
+// based on the given watch events and process batch function.
+func newEventQueueConsumerFunc(watchEvents []string, processBatchFunc ProcessBatchFunc) EventConsumerFunc {
+	return func(ctx *api.Context, batchSize int) error {
```

Comment:
Whats the advantage of returning this as a function vs calling the function directly ?

Reply:
Is it for passing a different consumer func to playbook runs?

Reply:
Ya, the `EventConsumer` needed to be generic and `watchEvents` and `processBatchFunc` are specific to consumers of `event_queue` and doesn't fit well with playbook even consumer.